CONTRIBUTING.md

```diff
@@ -0,0 +1,261 @@
+
+4. Favor reproducibilty
+
+    Always use fully resolved lockfiles or similar input files to determine what content needs to be download for a specific project (e.g. npm's `package-lock.json`, a `pip-compiled` generated `requirements.txt`, etc). Resolving the dependencies during the prefetch will prevent its behavior from being deterministic, or in other words, the same repository and the same commit hash should always result in identical prefetch results.
+
+## Development
+
+### Virtual environment
+
+Set up a virtual environment that has everything you will need for development:
+
+```shell
+make venv
+source venv/bin/activate
+```
+
+This installs the Cachi2 CLI in [editable mode](https://setuptools.pypa.io/en/latest/userguide/development_mode.html),
+which means changes to the source code will reflect in the behavior of the CLI without the need for reinstalling.
+
+You may need to install Python 3.9 in case you want to test your changes against Python 3.9 locally
+before submitting a pull request.
+
+```shell
+dnf install python3.9
+```
+
+The CLI also depends on the following non-Python dependencies:
+
+```shell
+dnf install golang-bin git
+```
+
+You should now have everything needed to [try out](#basic-usage) the CLI or hack on the code in ~~vim~~ your favorite
+editor.
+
+### Developer flags
+
+* `--dev-package-managers` (hidden): enables in-development package manager(s)
+  for test. Please refer to other existing package managers to see how they're
+  enabled and wired to the CLI.
+
+  Invoke it as `cachi2 fetch-deps --dev-package-managers FOO`
+
+  More explicitly
+
+  * `--dev-package-managers` is a *flag for* `fetch-deps`
+  * `FOO` is an *argument to* `fetch-deps` (i.e. the language to fetch for)
+
+### Coding standards
+
+Cachi2's codebase conforms to standards enforced by a collection of formatters, linters and other code checkers:
+
+* [black](https://black.readthedocs.io/en/stable/) (with a line-length of 100) for consistent formatting
+* [isort](https://pycqa.github.io/isort/) to keep imports sorted
+* [flake8](https://flake8.pycqa.org/en/latest/) to (de-)lint the code and ~~politely~~ ask for docstrings
+* [mypy](https://mypy.readthedocs.io/en/stable/) for type-checking. Please include type annotations for new code.
+* [pytest](https://docs.pytest.org/en/7.1.x/) to run unit tests and report coverage stats. Please aim for (near) full
+  coverage of new code.
+
+Options for all the tools are configured in [pyproject.toml](./pyproject.toml) and [tox.ini](./tox.ini).
+
+Run all the checks that your pull request will be subjected to:
+
+```shell
+make test
+```
```

Comment:
Basing on precedents I suggest adding this:

""" Please avoid modifying existing test cases unless fixing a bug that affects testing. If you extend existing feature please add a new test case instead of modifying any existing test cases. (It is fine to add a new parameter group though as long as test function remains unchanged)"""

Reply:
I see the point you're trying to make here, but I wonder if this'll be taken to an extreme? 

If a feature is changed, then the existing test cases will also change (which a different scenario from extending a feature, but I wonder if the line might get too blurry in some cases). 

And in some cases, even when extending, it might make sense to make small tweaks to an existing test (emphasis on small).

Wdyt? Should we try to rephrase this? Also, where do you think it fits? A small "how to write tests" section?

Reply:
What's an extreme here? Having separate test cases for separate features does not look like an extreme to me. Tests are code, but a different type of code, not like other code in codebase. They don't have to be non-repetitive, their two main goals are to test an actual scenario and to be easy to comprehend. Tests are not supposed to be read often and are not supposed to be modified unless a feature they are testing is getting redesigned or removed. However when one reads a test they better spend as little time as possible to figure out how a test works and what it tests. I think it is perfectly ok to copy a test implementation over, tweak it for a new use case and have a new test that way. On the other hand extending existing tests with even more functionality makes them confusing and prone to bugs. When a test function accepts a single parameterized argument it is trivial to figure out how the most recent change broke it. It becomes a separate quest when there are many such arguments and when a test function is a sequence of set-ups, intermediate assertions and branching on parameters.

Reply:
> are not supposed to be modified unless a feature they are testing is getting redesigned or removed

Yeah, that's the point I was missing. The first sentence gives the impression that only bugfixes should change the current tests. I'd even say "changed" instead of "redesigned" ^. If you change how a feature works, then the tests should be updated, otherwise they'll just fail. 

Maybe we can write it as:

"""
When extending an existing feature, please add a new test case instead of modifying any existing ones. Large test scenarios with many branching paths are very hard to understand and to maintain. It is ok to copy and paste large parts of an existing test if needed for a new scenario. Also, it is fine to add a new parameter group, though as long as the test function remains unchanged.
"""

Reply:
Yes, your suggestion reads better. I would have dropped 'though' from the last sentence, but that's an insignificant nitpick.

Reply:
Added it as a separate section (I followed this [logic](https://github.com/containerbuildsystem/cachi2/pull/649#discussion_r1769346092))